### ISSUE TITLE
[BUG] Fma bug in linux

### ISSUE THREAD

**Issue by andresnowak** [modularml/mojo#3185]
### Bug description

There seems to be a bug in FMA op when using float16 and bfloat16 values

### Steps to reproduce

Code: https://github.com/Benny-Nottonson/Mojo-Marathons/tree/cd3c6f2ba39500ba94736312b55fef36a4c4c4d6.
When running main.mojo and running the test in linux I get this error `assertionError: 24.4375 is not close to 24.453125 with a diff of 0.015625 mojo: error: execution exited with a non-zero result: 1` for the (1,47,97) matrix shape, but when using float32 or float64 or changing the basic_matmul in test.mojo to not use the fma operator or running it in MacOs the test runs correctly

### System information

```shell
- What OS did you do install Mojo on ? PopOs And MacOs
- Provide version information for Mojo by pasting the output of `mojo -v` 24.4.0
- Provide Modular CLI version by pasting the output of `modular -v` 0.8.0
```

**Comment by martinvuyk** [comment]
I think this is because your CPU has no hardware FMA and whatever math library runs on your system emulates it and has precision differences. see: [this stackoverflow answer](https://stackoverflow.com/a/28631091)

**Comment by andresnowak** [comment]
I have a ryzen 3600x and it seems it does have the operator so I don't think that is the problem.

**Comment by martinvuyk** [comment]
Then it is that the tolerance is too low, `24.4375 is not close to 24.453125 with a diff of 0.015625` is close depending on the application. And there will always be differences when doing fma against normal multiply and add since fma does 1 rounding and multiply and add do 1 each.

Excerpt from this [stackoverflow answer](https://stackoverflow.com/a/62273547):

Additionally, there's a whole other can of worms about the slight differences in the results from std::fma and (a*b)+c due to the way rounding is handled with floating point numbers. std::fma only rounds once during the two floating point operations, while (a*b)+c might[1] do a*b, store the result in 64 bits, add c to this value and then store the result in 64 bits.

If you want to minimize floating point arithmetic error in your calculations, std::fma is probably a better choice because it guarantees you will only have precious bits stripped away from your precious floating point numbers once.

[1] Whether or not this extra rounding happens depends on your compiler, your optimization settings and your architecture settings: [Compiler Explorer](https://godbolt.org/z/KXvAQN) examples for msvc, gcc, icc, clang

**Comment by andresnowak** [comment]
Hmm okay, but the thing is this problem only happens with float 16, like I think a 1e-2 error is already big. And also when using fma with float 16 the operations are a lot slower, that's what was also strange. So that's why I was thinking it is something else.

**Comment by martinvuyk** [comment]
It could also be that your CPU only has FP32 and/or FP64 units and is converting the FP16 before calculating 

Found another [stackoverflow answer](https://stackoverflow.com/a/4584707):
- If the hardware implements only double, then float will be slower if conversion to/from the native double format isn't free as part of float-load and float-store instructions.
- If the hardware implements float only, then emulating double with it will cost even more time. In this case, float will be faster.
- And if the hardware implements neither, and both have to be implemented in software. In this case, both will be slow, but double will be slightly slower (more load and store operations at the least).

I think this is getting into the realm of stuff the compiler really should resolve according to the target microarchitecture

**Comment by andresnowak** [comment]
Yeah that's why I made the issue, like Maybe this is what is happening but I think the compiler should solve the problem depending as you say for the architecture is compiling

**Comment by martinvuyk** [comment]
> float and double

float = FP32 and double = FP64
FP16 is pretty new and niche to Neural Networks so this is likely a problem with them assuming users have a newer chip. I also have a pretty old CPU and fma is slow as a snail but in my case I have no FMA at all

> compiler should solve the problem

totally, maybe change the title of the issue to something like "FMA instruction lowering issue, slow execution for architectures with no hardware IC"

**Comment by JoeLoser** [comment]
Is this still an issue?

**Comment by andresnowak** [comment]
For now i don't have access to this computer, so I wouldn't be able to confirm if it still exists. But this bug was on a ryzen 3600x using PopOs linux (if it helps to replicate the problem).
